### ISSUE TITLE
CUDA no longer compatible with Intel Clang

### ISSUE THREAD

**Issue by krasznaa** [intel/llvm#8037]
# Describe the bug

This is not necessarily a bug in `intel/llvm`, or a bug at all. But since I only recently ran into it, I thought it would be worth discussing about it a bit. (@ivorobts, @fwyzard. this is the issue that I mentioned in our meeting today.)

Recent versions of the Intel `clang++` compiler can not be used as the "host compiler" of `nvcc` / CUDA. :frowning: For the longest time one had to use the `-allow-unsupported-compiler` compiler flag with `nvcc` to make it work with the Intel compiler as the "host compiler". But with that, it worked. Unfortunately with the latest `intel/llvm` versions it no longer does. :frowning:

# To Reproduce

One simply needs some CUDA code first of all. Let's say:

```cuda
// CUDA include(s).
#include <cuda.h>

// System include(s).
#include <cstdio>

__global__ void myKernel() {

   printf("Hello World!\n");
}

int main() {

   myKernel<<<1, 1>>>();
   cudaDeviceSynchronize();
   return 0;
}
```

Then, one needs to set up the Intel compiler, and CUDA for the runtime environment. With those in place, the following compilation command used to work:

```
[bash][pcadp04]:oneAPI-CUDA-clash > source /software/intel/clang-2022-09/x86_64-centos9-gcc11-opt/setup.sh 
Configured CUDA from: /software/cuda/11.7.1/x86_64
Configured Clang from: /software/intel/clang-2022-09/x86_64-centos9-gcc11-opt
[bash][pcadp04]:oneAPI-CUDA-clash > nvcc -ccbin=`which clang++` -allow-unsupported-compiler ./helloWorld.cu 
[bash][pcadp04]:oneAPI-CUDA-clash > ./a.out 
Hello World!
[bash][pcadp04]:oneAPI-CUDA-clash >
```

With the latest nightly, I rather get the following:

```
[bash][pcadp04]:oneAPI-CUDA-clash > source /software/intel/nightly/20230117/x86_64-centos9-gcc11-opt/setup.sh 
Configured CUDA from: /software/cuda/12.0.0/x86_64
Configured Clang from: /software/intel/nightly/20230117/x86_64-centos9-gcc11-opt
[bash][pcadp04]:oneAPI-CUDA-clash > nvcc -ccbin=`which clang++` -allow-unsupported-compiler ./helloWorld.cu
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/ext/type_traits.h(193): error: parameter pack "_Tp" was referenced but not expanded

/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/ext/type_traits.h(193): error: expected an expression

/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/type_traits(3003): error: parameter pack "_ArgTypes" was referenced but not expanded

/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/type_traits(3003): error: expected an expression

/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/type_traits(3019): error: parameter pack "_ArgTypes" was referenced but not expanded

/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/type_traits(3019): error: expected an expression

/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/type_traits(3031): error: parameter pack "_ArgTypes" was referenced but not expanded

/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/type_traits(3031): error: expected an expression

/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/type_traits(3046): error: parameter pack "_ArgTypes" was referenced but not expanded

/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/type_traits(3046): error: expected an expression

/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/type_traits(3071): error: parameter pack "_ArgTypes" was referenced but not expanded

/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/type_traits(3071): error: expected an expression

/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/bits/stl_pair.h(460): error: argument list for class template "std::pair" is missing

/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/bits/stl_pair.h(460): error: expected a ")"

/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/bits/stl_pair.h(460): error: expected a ";"

15 errors detected in the compilation of "./helloWorld.cu".
[bash][pcadp04]:oneAPI-CUDA-clash >
```

GCC 11 is not special in the output. On Ubuntu 20.04, with GCC 9, I still see issues:

```
[bash][atspot01]:oneAPI-CUDA-clash > source /software/intel/oneapi-2023.0.0/setvars.sh 
 
:: initializing oneAPI environment ...
   bash: BASH_VERSION = 5.0.17(1)-release
   args: Using "$@" for setvars.sh arguments: 
:: advisor -- latest
:: ccl -- latest
:: compiler -- latest
:: dal -- latest
:: debugger -- latest
:: dev-utilities -- latest
:: dnnl -- latest
:: dpcpp-ct -- latest
:: dpl -- latest
:: ipp -- latest
:: ippcp -- latest
:: ipp -- latest
:: mkl -- latest
:: mpi -- latest
:: tbb -- latest
:: vpl -- latest
:: vtune -- latest
:: oneAPI environment initialized ::
 
[bash][atspot01]:oneAPI-CUDA-clash > source /software/cuda/12.0.0/x86_64/setup.sh 
Configured CUDA from: /software/cuda/12.0.0/x86_64
[bash][atspot01]:oneAPI-CUDA-clash > nvcc -ccbin=`which icpx` -allow-unsupported-compiler ./helloWorld.cu/usr/lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/stl_pair.h(442): error: argument list for class template "std::pair" is missing

/usr/lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/stl_pair.h(442): error: expected a ")"

/usr/lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/stl_pair.h(442): error: expected a ";"

3 errors detected in the compilation of "./helloWorld.cu".
[bash][atspot01]:oneAPI-CUDA-clash >
```

While once again, with earlier versions this worked fine:

```
[bash][atspot01]:oneAPI-CUDA-clash > source /software/intel/oneapi-2022.3.1/setvars.sh 
 
:: initializing oneAPI environment ...
   bash: BASH_VERSION = 5.0.17(1)-release
   args: Using "$@" for setvars.sh arguments: 
:: advisor -- latest
:: ccl -- latest
:: compiler -- latest
:: dal -- latest
:: debugger -- latest
:: dev-utilities -- latest
:: dnnl -- latest
:: dpcpp-ct -- latest
:: dpl -- latest
:: intelpython -- latest
:: ipp -- latest
:: ippcp -- latest
:: ipp -- latest
:: mkl -- latest
:: mpi -- latest
:: tbb -- latest
:: vpl -- latest
:: vtune -- latest
:: oneAPI environment initialized ::
 
[bash][atspot01]:oneAPI-CUDA-clash > source /software/cuda/12.0.0/x86_64/setup.sh 
Configured CUDA from: /software/cuda/12.0.0/x86_64
[bash][atspot01]:oneAPI-CUDA-clash > nvcc -ccbin=`which icpx` -allow-unsupported-compiler ./helloWorld.cu
[bash][atspot01]:oneAPI-CUDA-clash > ./a.out 
Hello World!
[bash][atspot01]:oneAPI-CUDA-clash >
```

# Environment (please complete the following information):

- OS: CentOS Stream 9, Ubuntu 20.04, Ubuntu 22.04 (I don't think the exact version makes a difference. These are the ones that I actually tried.)
- Target device and vendor: N/A
- DPC++ version:

```
Intel(R) oneAPI DPC++/C++ Compiler 2023.0.0 (2023.0.0.20221201)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /mnt/hdd1/software/intel/oneapi-2023.0.0/compiler/2023.0.0/linux/bin-llvm
Configuration file: /mnt/hdd1/software/intel/oneapi-2023.0.0/compiler/2023.0.0/linux/bin-llvm/../bin/icpx.cfg
Found candidate GCC installation: /usr/lib/gcc/x86_64-redhat-linux/11
Selected GCC installation: /usr/lib/gcc/x86_64-redhat-linux/11
```

```
clang version 16.0.0 (https://github.com/intel/llvm.git baa3d73963a27d90d38e966afb83831633562d9a)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /software/intel/nightly/20230117/x86_64-centos9-gcc11-opt/bin
Found candidate GCC installation: /usr/lib/gcc/x86_64-redhat-linux/11
Selected GCC installation: /usr/lib/gcc/x86_64-redhat-linux/11
```

- Dependencies version: N/A

# Additional context

Some of my first experiences with this were documented in: https://gitlab.kitware.com/cmake/cmake/-/issues/24320 As at first I thought that CMake was somehow at fault with this.

**Comment by jle-quel** [comment]
Hello :wave:

I've been looking at this issue, and I've been able to reproduce the error.

This "feature" has been broken since clang's default C++ dialect has been updated to `gnu++17`.
Here's the relevant commit: "https://github.com/intel/llvm/commit/3e99b8d947ac024831e59be2b604ac67a24fed94"

TL;DR

Two solutions:
- Invoke `nvcc` by passing the option `-std=c++17`, which will set the c++'s dialect version for `nvcc`'s toolchain to match the host compiler.
- Invoke `nvcc` by passing the options `-Xcompiler -std=gnu++14`, which will set the host compiler c++'s dialect version to match `nvcc`'s toolchain.


Here's the summary of this investigation:

Looking at the `nvcc` documentation, we can see these two phrases:
- "The option "--std" select a particular C++ dialect. Note that this flag also turns on the corresponding dialect flag for the host compiler."
- "The default C++ dialect depends on the host compiler. `nvcc` matches the default C++ dialect that the host compiler uses."

The problem comes from "nvcc matches the default C++ dialect that the host compiler uses".

I discovered that the CUDA compiler driver does not propagate the C++ dialect to its toolchain.
In other words, this error is not specific to dpc++ and will generate equivalent
errors, if you use a host compiler which have it's C++ dialect set to 17 and/or greater

Here's this actions created by `nvcc` when compiling a `.cu` file:

```
> nvcc -dryrun -ccbin=$LLVM_BUILD/clang++ -allow-unsupported-compiler main.cu

#$ "llvm/build/bin"/clang++ -D__NV_NO_HOST_COMPILER_CHECK=1 -D__CUDA_ARCH__=520 -D__CUDA_ARCH_LIST__=520 -E -x c++  -DCUDA_DOUBLE_MATH_FUNCTIONS -D__CUDACC__ -D__NVCC__  "-I/usr/local/cuda-11.7/bin/../targets/x86_64-linux/include"    -D__CUDACC_VER_MAJOR__=11 -D__CUDACC_VER_MINOR__=7 -D__CUDACC_VER_BUILD__=99 -D__CUDA_API_VER_MAJOR__=11 -D__CUDA_API_VER_MINOR__=7 -D__NVCC_DIAG_PRAGMA_SUPPORT__=1 -include "cuda_runtime.h" -m64 "main.cu" -o "/tmp/tmpxft_00069888_00000000-9_main.cpp1.ii"
#$ cicc --c++14 --clang --clang_version=160000 --display_error_number --orig_src_file_name "main.cu" --orig_src_path_name "/home/jle-quel/Documents/cpp/samples/cuda/main.cu" --unicode_source_kind=UTF-8 --allow_managed   -arch compute_52 -m64 --no-version-ident -ftz=0 -prec_div=1 -prec_sqrt=1 -fmad=1 --include_file_name "tmpxft_00069888_00000000-3_main.fatbin.c" -tused --gen_module_id_file --module_id_file_name "/tmp/tmpxft_00069888_00000000-4_main.module_id" --gen_c_file_name "/tmp/tmpxft_00069888_00000000-6_main.cudafe1.c" --stub_file_name "/tmp/tmpxft_00069888_00000000-6_main.cudafe1.stub.c" --gen_device_file_name "/tmp/tmpxft_00069888_00000000-6_main.cudafe1.gpu"  "/tmp/tmpxft_00069888_00000000-9_main.cpp1.ii" -o "/tmp/tmpxft_00069888_00000000-6_main.ptx"
#$ ptxas -arch=sm_52 -m64  "/tmp/tmpxft_00069888_00000000-6_main.ptx"  -o "/tmp/tmpxft_00069888_00000000-10_main.sm_52.cubin"
#$ fatbinary -64 --cicc-cmdline="-ftz=0 -prec_div=1 -prec_sqrt=1 -fmad=1 " "--image3=kind=elf,sm=52,file=/tmp/tmpxft_00069888_00000000-10_main.sm_52.cubin" "--image3=kind=ptx,sm=52,file=/tmp/tmpxft_00069888_00000000-6_main.ptx" --embedded-fatbin="/tmp/tmpxft_00069888_00000000-3_main.fatbin.c"
#$ rm /tmp/tmpxft_00069888_00000000-3_main.fatbin
#$ "llvm/build/bin"/clang++ -D__NV_NO_HOST_COMPILER_CHECK=1 -D__CUDA_ARCH_LIST__=520 -E -x c++ -D__CUDACC__ -D__NVCC__  "-I/usr/local/cuda-11.7/bin/../targets/x86_64-linux/include"    -D__CUDACC_VER_MAJOR__=11 -D__CUDACC_VER_MINOR__=7 -D__CUDACC_VER_BUILD__=99 -D__CUDA_API_VER_MAJOR__=11 -D__CUDA_API_VER_MINOR__=7 -D__NVCC_DIAG_PRAGMA_SUPPORT__=1 -include "cuda_runtime.h" -m64 "main.cu" -o "/tmp/tmpxft_00069888_00000000-5_main.cpp4.ii"
#$ cudafe++ --c++14 --clang --clang_version=160000 --display_error_number --orig_src_file_name "main.cu" --orig_src_path_name "/home/jle-quel/Documents/cpp/samples/cuda/main.cu" --unicode_source_kind=UTF-8 --allow_managed  --m64 --parse_templates --gen_c_file_name "/tmp/tmpxft_00069888_00000000-6_main.cudafe1.cpp" --stub_file_name "tmpxft_00069888_00000000-6_main.cudafe1.stub.c" --module_id_file_name "/tmp/tmpxft_00069888_00000000-4_main.module_id" "/tmp/tmpxft_00069888_00000000-5_main.cpp4.ii"
#$ "llvm/build/bin"/clang++ -D__CUDA_ARCH__=520 -D__CUDA_ARCH_LIST__=520 -c -x c++  -DCUDA_DOUBLE_MATH_FUNCTIONS "-I/usr/local/cuda-11.7/bin/../targets/x86_64-linux/include"   -m64 "/tmp/tmpxft_00069888_00000000-6_main.cudafe1.cpp" -o "/tmp/tmpxft_00069888_00000000-11_main.o"
#$ nvlink -m64 --arch=sm_52 --register-link-binaries="/tmp/tmpxft_00069888_00000000-7_a_dlink.reg.c"    "-L/usr/local/cuda-11.7/bin/../targets/x86_64-linux/lib/stubs" "-L/usr/local/cuda-11.7/bin/../targets/x86_64-linux/lib" -cpu-arch=X86_64 "/tmp/tmpxft_00069888_00000000-11_main.o"  -lcudadevrt  -o "/tmp/tmpxft_00069888_00000000-12_a_dlink.sm_52.cubin" --host-ccbin "llvm/build/bin/clang++"
#$ fatbinary -64 --cicc-cmdline="-ftz=0 -prec_div=1 -prec_sqrt=1 -fmad=1 " -link "--image3=kind=elf,sm=52,file=/tmp/tmpxft_00069888_00000000-12_a_dlink.sm_52.cubin" --embedded-fatbin="/tmp/tmpxft_00069888_00000000-8_a_dlink.fatbin.c"
#$ rm /tmp/tmpxft_00069888_00000000-8_a_dlink.fatbin
#$ "llvm/build/bin"/clang++ -D__CUDA_ARCH_LIST__=520 -c -x c++ -DFATBINFILE="\"/tmp/tmpxft_00069888_00000000-8_a_dlink.fatbin.c\"" -DREGISTERLINKBINARYFILE="\"/tmp/tmpxft_00069888_00000000-7_a_dlink.reg.c\"" -I. -D__NV_EXTRA_INITIALIZATION= -D__NV_EXTRA_FINALIZATION= -D__CUDA_INCLUDE_COMPILER_INTERNAL_HEADERS__  "-I/usr/local/cuda-11.7/bin/../targets/x86_64-linux/include"    -D__CUDACC_VER_MAJOR__=11 -D__CUDACC_VER_MINOR__=7 -D__CUDACC_VER_BUILD__=99 -D__CUDA_API_VER_MAJOR__=11 -D__CUDA_API_VER_MINOR__=7 -D__NVCC_DIAG_PRAGMA_SUPPORT__=1 -m64 "/usr/local/cuda-11.7/bin/crt/link.stub" -o "/tmp/tmpxft_00069888_00000000-13_a_dlink.o"
#$ "llvm/build/bin"/clang++ -D__CUDA_ARCH_LIST__=520 -m64 -Wl,--start-group "/tmp/tmpxft_00069888_00000000-13_a_dlink.o" "/tmp/tmpxft_00069888_00000000-11_main.o"   "-L/usr/local/cuda-11.7/bin/../targets/x86_64-linux/lib/stubs" "-L/usr/local/cuda-11.7/bin/../targets/x86_64-linux/lib"  -lcudadevrt  -lcudart_static  -lrt -lpthread  -ldl  -Wl,--end-group -o "a.out"

```

As you can see, the binary `cicc` and `cudafe++` have the option to set the c++ dialect to 14, whereas the default for clang is now 17.
Until the commit, which changed clang C++ dialect to 17, this worked correctly since the driver's toolchain used 14.

But since the update to 17, we can see that `nvcc` does not match the default C++ dialect that the host compiler uses.

Finally, here's the action created by `nvcc` when compiling a `.cu` file, and by explicitly specifying the C++ dialect, so that `nvcc` can propagate it to its toolchain.

```
> nvcc -dryrun -std=c++17 -ccbin=$LLVM_BUILD/clang++ -allow-unsupported-compiler main.cu

#$ "llvm/build/bin"/clang++ -D__NV_NO_HOST_COMPILER_CHECK=1 -std=c++17 -D__CUDA_ARCH__=520 -D__CUDA_ARCH_LIST__=520 -E -x c++  -DCUDA_DOUBLE_MATH_FUNCTIONS -D__CUDACC__ -D__NVCC__  "-I/usr/local/cuda-11.7/bin/../targets/x86_64-linux/include"    -D__CUDACC_VER_MAJOR__=11 -D__CUDACC_VER_MINOR__=7 -D__CUDACC_VER_BUILD__=99 -D__CUDA_API_VER_MAJOR__=11 -D__CUDA_API_VER_MINOR__=7 -D__NVCC_DIAG_PRAGMA_SUPPORT__=1 -include "cuda_runtime.h" -m64 "main.cu" -o "/tmp/tmpxft_00069a39_00000000-9_main.cpp1.ii"
#$ cicc --c++17 --clang --clang_version=160000 --display_error_number --orig_src_file_name "main.cu" --orig_src_path_name "/home/jle-quel/Documents/cpp/samples/cuda/main.cu" --unicode_source_kind=UTF-8 --allow_managed   -arch compute_52 -m64 --no-version-ident -ftz=0 -prec_div=1 -prec_sqrt=1 -fmad=1 --include_file_name "tmpxft_00069a39_00000000-3_main.fatbin.c" -tused --gen_module_id_file --module_id_file_name "/tmp/tmpxft_00069a39_00000000-4_main.module_id" --gen_c_file_name "/tmp/tmpxft_00069a39_00000000-6_main.cudafe1.c" --stub_file_name "/tmp/tmpxft_00069a39_00000000-6_main.cudafe1.stub.c" --gen_device_file_name "/tmp/tmpxft_00069a39_00000000-6_main.cudafe1.gpu"  "/tmp/tmpxft_00069a39_00000000-9_main.cpp1.ii" -o "/tmp/tmpxft_00069a39_00000000-6_main.ptx"
#$ ptxas -arch=sm_52 -m64  "/tmp/tmpxft_00069a39_00000000-6_main.ptx"  -o "/tmp/tmpxft_00069a39_00000000-10_main.sm_52.cubin"
#$ fatbinary -64 --cicc-cmdline="-ftz=0 -prec_div=1 -prec_sqrt=1 -fmad=1 " "--image3=kind=elf,sm=52,file=/tmp/tmpxft_00069a39_00000000-10_main.sm_52.cubin" "--image3=kind=ptx,sm=52,file=/tmp/tmpxft_00069a39_00000000-6_main.ptx" --embedded-fatbin="/tmp/tmpxft_00069a39_00000000-3_main.fatbin.c"
#$ rm /tmp/tmpxft_00069a39_00000000-3_main.fatbin
#$ "llvm/build/bin"/clang++ -D__NV_NO_HOST_COMPILER_CHECK=1 -std=c++17 -D__CUDA_ARCH_LIST__=520 -E -x c++ -D__CUDACC__ -D__NVCC__  "-I/usr/local/cuda-11.7/bin/../targets/x86_64-linux/include"    -D__CUDACC_VER_MAJOR__=11 -D__CUDACC_VER_MINOR__=7 -D__CUDACC_VER_BUILD__=99 -D__CUDA_API_VER_MAJOR__=11 -D__CUDA_API_VER_MINOR__=7 -D__NVCC_DIAG_PRAGMA_SUPPORT__=1 -include "cuda_runtime.h" -m64 "main.cu" -o "/tmp/tmpxft_00069a39_00000000-5_main.cpp4.ii"
#$ cudafe++ --c++17 --clang --clang_version=160000 --display_error_number --orig_src_file_name "main.cu" --orig_src_path_name "/home/jle-quel/Documents/cpp/samples/cuda/main.cu" --unicode_source_kind=UTF-8 --allow_managed  --m64 --parse_templates --gen_c_file_name "/tmp/tmpxft_00069a39_00000000-6_main.cudafe1.cpp" --stub_file_name "tmpxft_00069a39_00000000-6_main.cudafe1.stub.c" --module_id_file_name "/tmp/tmpxft_00069a39_00000000-4_main.module_id" "/tmp/tmpxft_00069a39_00000000-5_main.cpp4.ii"
#$ "llvm/build/bin"/clang++ -std=c++17 -D__CUDA_ARCH__=520 -D__CUDA_ARCH_LIST__=520 -c -x c++  -DCUDA_DOUBLE_MATH_FUNCTIONS "-I/usr/local/cuda-11.7/bin/../targets/x86_64-linux/include"   -m64 "/tmp/tmpxft_00069a39_00000000-6_main.cudafe1.cpp" -o "/tmp/tmpxft_00069a39_00000000-11_main.o"
#$ nvlink -m64 --arch=sm_52 --register-link-binaries="/tmp/tmpxft_00069a39_00000000-7_a_dlink.reg.c"    "-L/usr/local/cuda-11.7/bin/../targets/x86_64-linux/lib/stubs" "-L/usr/local/cuda-11.7/bin/../targets/x86_64-linux/lib" -cpu-arch=X86_64 "/tmp/tmpxft_00069a39_00000000-11_main.o"  -lcudadevrt  -o "/tmp/tmpxft_00069a39_00000000-12_a_dlink.sm_52.cubin" --host-ccbin "llvm/build/bin/clang++"
#$ fatbinary -64 --cicc-cmdline="-ftz=0 -prec_div=1 -prec_sqrt=1 -fmad=1 " -link "--image3=kind=elf,sm=52,file=/tmp/tmpxft_00069a39_00000000-12_a_dlink.sm_52.cubin" --embedded-fatbin="/tmp/tmpxft_00069a39_00000000-8_a_dlink.fatbin.c"
#$ rm /tmp/tmpxft_00069a39_00000000-8_a_dlink.fatbin
#$ "llvm/build/bin"/clang++ -std=c++17 -D__CUDA_ARCH_LIST__=520 -c -x c++ -DFATBINFILE="\"/tmp/tmpxft_00069a39_00000000-8_a_dlink.fatbin.c\"" -DREGISTERLINKBINARYFILE="\"/tmp/tmpxft_00069a39_00000000-7_a_dlink.reg.c\"" -I. -D__NV_EXTRA_INITIALIZATION= -D__NV_EXTRA_FINALIZATION= -D__CUDA_INCLUDE_COMPILER_INTERNAL_HEADERS__  "-I/usr/local/cuda-11.7/bin/../targets/x86_64-linux/include"    -D__CUDACC_VER_MAJOR__=11 -D__CUDACC_VER_MINOR__=7 -D__CUDACC_VER_BUILD__=99 -D__CUDA_API_VER_MAJOR__=11 -D__CUDA_API_VER_MINOR__=7 -D__NVCC_DIAG_PRAGMA_SUPPORT__=1 -m64 "/usr/local/cuda-11.7/bin/crt/link.stub" -o "/tmp/tmpxft_00069a39_00000000-13_a_dlink.o"
#$ "llvm/build/bin"/clang++ -D__CUDA_ARCH_LIST__=520 -m64 -std=c++17 -Wl,--start-group "/tmp/tmpxft_00069a39_00000000-13_a_dlink.o" "/tmp/tmpxft_00069a39_00000000-11_main.o"   "-L/usr/local/cuda-11.7/bin/../targets/x86_64-linux/lib/stubs" "-L/usr/local/cuda-11.7/bin/../targets/x86_64-linux/lib"  -lcudadevrt  -lcudart_static  -lrt -lpthread  -ldl  -Wl,--end-group -o "a.out"
```

To conclude, to be able to use Intel clang++ compiler as the "host compiler" for `nvcc/CUDA`,
you will need to either explicitly pass the C++ dialect to `nvcc` or set the clang C++
dialect to a lower version than its default since `nvcc` does not matches the default C++ dialect that the host compiler uses

**Comment by krasznaa** [comment]
This is **super useful** @jle-quel! I'll give this a try. But it sounds very much like it should allow me to make use of CUDA alongside the latest Intel compilers. :smile:

**Comment by jle-quel** [comment]
Let me know if you encounter any more problems. Otherwise, I'll let you close the issue :smile:

**Comment by krasznaa** [comment]
Yepp, things work fine in my build as long as I use `CUDAFLAGS="-std=c++17 ..."` in my environment with my CMake based build. This is not really a robust setup, as it requires the environment setup and CMake cache variable settings to be in sync for a successful build, but it should be fine for my developments for the moment.

I'll follow up with Kitware and NVIDIA on:
  - https://gitlab.kitware.com/cmake/cmake/-/issues/24320
  - https://forums.developer.nvidia.com/t/using-nvcc-with-the-latest-version-of-llvm-clang

Thanks again for your investigation!
